### PR TITLE
Bring back SKIPPED action logic, and handling run cancellation in the Runner.

### PR DESCRIPTION
**Pull request by BenBirt**
for dataform-co/dataform-co#1968

### CODE REVIEW

File: api/commands/run.ts [hunk 1]
```diff
@@ -112,24 +112,54 @@ export class Runner {
...
-        this.result.actions.push(await this.executeAction(action));
-        await this.triggerChange();
-        await this.executeAllActionsReadyForExecution();
-      })
+    // If the run has been cancelled, skip all pending actions.
+    if (this.cancelled) {
+      const allPendingActions = this.pendingActions;
+      this.pendingActions = [];
+      await Promise.all(
+        allPendingActions.map(async pendingAction => this.skipAction(pendingAction))
```
Comment (lewish) [inline api/commands/run.ts]: These aren't promises, no need to await or promise.all() here. `skipAction` implies it mutates state, but it does not - `skippedAction` or `getSkippedAction` ? Or could just be inline tbh...

File: api/commands/run.ts [hunk 2]
```diff
@@ -112,24 +112,54 @@ export class Runner {
...
+          this.onActionExecutionComplete(await this.executeAction(executableAction))
+        )
+      )
+    ]);
+  }
+
+  private async onActionExecutionComplete(executedAction: dataform.IExecutedAction) {
+    this.result.actions.push(executedAction);
+    await this.triggerChange();
+    await this.executeAllActionsReadyForExecution();
```
Comment (lewish) [inline api/commands/run.ts]: Hmmm. AIUI, you build up this massive dependency tree of promises that's the inverse of the dependecy graph, that's going to recursively come all the way back up to the top when it's done?
Feels like when this goes wrong it's going to be very hard to debug / reason about.
Personally I thought the for loop was more readable. (Not going to block on this though).

File: api/commands/run.ts (hunk 1 above)
Comment (BenBirt) [reply api/commands/run.ts]: fixed, i had accidentally not called `onActionExecutionComplete` on these (which does return a Promise)

File: api/commands/run.ts (hunk 2 above)
Comment (BenBirt) [reply api/commands/run.ts]: unless I misunderstood the previous code, this code is very similar to the loop, which was itself recursive (or at least, triggered calls to itself).

i could be misunderstanding your comment, but I don't think it works like that. this code does not build an inverse of the dependency graph, it's very like the old code in that upon each iteration it (a) finds and executes all executable actions, (b) skips all skippable actions, (c) triggers another iteration of all of the above, until the list of pending actions is exhausted.

**Pull request merged**